### DISCUSSION TITLE
How do I start the ssh-agent?

### DISCUSSION THREAD

**Discussion by theRoboxx** [leo-arch/clifm#202] [Q&A]
I tried 
```
$ eval `ssh-agent`
Agent pid 8020
```
to start the ssh-agent but after that I get the response
```
$ ssh-add -t 600 ~/.ssh/id_rsa
Could not open a connection to your authentication agent.
```
Whereas in bash and zsh `ssh-add` asks for my passphrase of the key.

Do I need to start the ssh-agent with a different command?

**Comment by leo-arch** [comment]
Hi @theRoboxx. I'll take a look at it and let you know.

Thanks for reporting.

**Comment by leo-arch** [comment]
The problem is the `export` line produced by `eval ssh-agent`. Basically, **SSH_AUTH_SOCK** and **SSH_AGENT_PID** are not being exported to the environment. Why?

1. `export` is a shell built-in, and as such is not recognized by **clifm**, which runs as a _child_ of the shell: it can run whatever executable file is in your **PATH**, but not shell internals (functions, aliases, built-in commands, and so on)
2. **Clifm**'s `export` function does not work the same as the shell built-in (which is why `eval` fails)

As a workaround:

```sh
# Run the ssh agent
ssh-agent

# Export the values outputted by the previous command (using clifm's built-in export function) 
export SSH_AUTH_SOCK=/tmp/ssh...
export SSH_AGENT_PID=...

# Now run ssh-add
ssh-add -t 600 ~/.ssh/id_rsa
```

**Comment by theRoboxx** [comment]
Thanks for your response.

I'm trying to automate this process by running a bash script.
Is there a way to run the clifm-export-function from within a bash script or do I need to execute the command manually in clifm?

**Reply by leo-arch** [reply]
A shell script is read by your shell, and thereby knows nothing about **clifm** internals.

It might work nonetheless as a plugin. Let me tinker with this idea.

**Reply by leo-arch** [reply]
Yes, it works, it's more like a workaround, but it works.

We just need to replace the first of your original commands by a plugin. Here's how.

1. Create a new plugin in `~/.config/clifm/plugins` named `start_ssh_agent.sh` with the following content:

```sh
#!/bin/sh

###
# Description: Start the SSH agent from within clifm
# Author: L. Abramovich
# License: GPL3
#
# Dependencies: openssh, cut, sed
###

vars="$(ssh-agent | cut -d';' -f1)"
var1="$(echo "$vars" | sed -n 1p)"
var2="$(echo "$vars" | sed -n 2p)"

printf "export %s; export %s" "$var1" "$var2" > "$CLIFM_BUS"
```
2. Make it executable
3. Create a new action (via `actions edit`) pointing to this script:
```
ssa=start_ssh_agent.sh
```

4. Run the plugin
```
ssa
```
5. Finally, you can add your SSH keys as usual
```
ssh-add -t 600 FILE
```

**Reply by leo-arch** [reply]
Hi @theRoboxx. Did it work?

**Reply by theRoboxx** [reply]
Yes, it worked perfectly.

I'm currently trying to extend the script such that I run it one time and every new shell also sets SSH_AUTH_SOCK and SSH_AGENT_PID correctly. 
`export` only works for the current clifm instance, so I need to start `ssh-agent` once but `export` every time, correct?

What is the easiest way to run a script at the beginning for every instance of clifm? Execute it in the clifmrc?

**Reply by leo-arch** [reply]
Adding it to your [profile file](https://github.com/leo-arch/clifm/wiki/Specifics#profile-file) is the right way.

**Reply by theRoboxx** [reply]
Quick question. Why can't I add the `ssh-add -t 600 FILE` command to the script of `ssa`?
```
#!/bin/sh

###
# Description: Start the SSH agent from within clifm
# Author: L. Abramovich
# License: GPL3
#
# Dependencies: openssh, cut, sed
###

vars="$(ssh-agent | cut -d';' -f1)"
var1="$(echo "$vars" | sed -n 1p)"
var2="$(echo "$vars" | sed -n 2p)"

printf "export %s; export %s" "$var1" "$var2" > "$CLIFM_BUS"


# only ask for the passphrase, if the key isn't unlocked
if ! [[ "$(ssh-add -l)" =~ "SHA" ]] ; then
  ssh-add -t 600 ~/.ssh/id_rsa
fi
```
It doesn't show the prompt to enter the passphrase. It's like the `printf "export %s; export %s" "$var1" "$var2" > "$CLIFM_BUS"` line wasn't executed.

But if I enter the following into  profile.clifm it works.

```
ssa

# only ask for the passphrase, if the key isn't unlocked
if ! [[ "$(ssh-add -l)" =~ "SHA" ]] ; then
  ssh-add -t 600 ~/.ssh/id_rsa
fi

```

**Reply by leo-arch** [reply]
Hey @theRoboxx. Yes, `printf "export %s; export %s" "$var1" "$var2" > "$CLIFM_BUS"` will be executed immediately after the script's end by **clifm**. This line just _writes_ the command to a temp file, which is read and executed _after_ the plugin exits.

**Reply by leo-arch** [reply]
I'd try adding the `ssh-add` command to the line sent to **CLIFM_BUS**, something like this:

```
printf "export %s; export %s" "$var1" "$var2; if ! [[ "$(ssh-add -l)" =~ "SHA" ]]; then ssh-add -t 600 ~/.ssh/id_rsa; fi" > "$CLIFM_BUS"

**Reply by theRoboxx** [reply]
Thanks for your reply.
Unfortunately this throws errors and I couldn't figure out why.

```
Could not open a connection to your authentication agent.
sh: -c: line 1: conditional binary operator expected
sh: -c: line 1: syntax error near `SHA'
sh: -c: line 1: `if ! [[  =~ SHA ]]'
sh: -c: line 1: syntax error near unexpected token `then'
sh: -c: line 1: `then ssh-add -t 600 /home/stew/.ssh/id_rsa'
sh: -c: line 1: syntax error near unexpected token `fi'
sh: -c: line 1: `fi'
```
Changing the brackets seems to eliminate two of the errors:
```
printf "export %s; export %s" "$var1" "$var2; if ! ( "$(ssh-add -l)" =~ "SHA" ); then ssh-add -t 600 ~/.ssh/id_rsa; fi" > "$CLIFM_BUS"`
```
```
Could not open a connection to your authentication agent.
sh: -c: line 2: syntax error: unexpected end of file
sh: -c: line 1: syntax error near unexpected token `then'
sh: -c: line 1: `then ssh-add -t 600 /home/stew/.ssh/id_rsa'
sh: -c: line 1: syntax error near unexpected token `fi'
sh: -c: line 1: `fi'
```
